server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/AbstractRareTermsAggregator.java

```diff
@@ -0,0 +1,141 @@
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.search.aggregations.bucket.terms;
+
+import org.elasticsearch.common.Randomness;
+import org.elasticsearch.common.util.SetBackedScalingCuckooFilter;
+import org.elasticsearch.search.DocValueFormat;
+import org.elasticsearch.search.aggregations.Aggregator;
+import org.elasticsearch.search.aggregations.AggregatorFactories;
+import org.elasticsearch.search.aggregations.BucketOrder;
+import org.elasticsearch.search.aggregations.LeafBucketCollector;
+import org.elasticsearch.search.aggregations.bucket.DeferableBucketAggregator;
+import org.elasticsearch.search.aggregations.bucket.DeferringBucketCollector;
+import org.elasticsearch.search.aggregations.bucket.MergingBucketsDeferringCollector;
+import org.elasticsearch.search.aggregations.bucket.nested.NestedAggregator;
+import org.elasticsearch.search.aggregations.pipeline.PipelineAggregator;
+import org.elasticsearch.search.aggregations.support.ValuesSource;
+import org.elasticsearch.search.internal.SearchContext;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+
+public abstract class AbstractRareTermsAggregator<T extends ValuesSource, U extends IncludeExclude.Filter>
+    extends DeferableBucketAggregator {
+
+    /**
+     Sets the number of "removed" values to accumulate before we purge ords
+     via the MergingBucketCollector's mergeBuckets() method
+     */
+    static final long GC_THRESHOLD = 1000000;
+    static final BucketOrder ORDER = BucketOrder.compound(BucketOrder.count(true), BucketOrder.key(true)); // sort by count ascending
+
+    protected final long maxDocCount;
+    protected final double precision;
+    protected final DocValueFormat format;
+    protected final T valuesSource;
+    protected final U includeExclude;
+
+    // Counter used during collection to track map entries that need GC'ing
+    long numDeleted = 0;
+
+    MergingBucketsDeferringCollector deferringCollector;
+    LeafBucketCollector subCollectors;
+    final SetBackedScalingCuckooFilter filter;
+
+    AbstractRareTermsAggregator(String name, AggregatorFactories factories, SearchContext context,
+                                Aggregator parent, List<PipelineAggregator> pipelineAggregators,
+                                Map<String, Object> metaData, long maxDocCount, double precision,
+                                DocValueFormat format, T valuesSource, U includeExclude) throws IOException {
+        super(name, factories, context, parent, pipelineAggregators, metaData);
+
+        this.filter = new SetBackedScalingCuckooFilter(10000, Randomness.get(), precision);
```

Comment:
One potential issue I see here is that the `Random` we get from `Randomness` will have a different seed each time meaning that multiple identical requests are not guaranteed to have the same response. I wonder if we should use a seed to overcome this? Maybe the index_shardID would work, or maybe there is something from the request we could use for a seed?

Reply:
Mmm, good point.  Will poke around